File: python3/sorting/merge_sort.py


def merge_sort(elements):
    if len(elements) > 1:
        mid = len(elements) // 2
        left_list = elements[:mid]
        right_list = elements[mid:]
        merge_sort(left_list)
        merge_sort(right_list)

        # merge

        i = 0
        j = 0
        k = 0

        while i < len(left_list) and j < len(right_list):

            if left_list[i] < right_list[j]:
                elements[k] = left_list[i]
                i += 1
                k += 1
            else:
                elements[k] = right_list[j]
                j += 1
                k += 1
        while i < len(left_list):
            elements[k] = left_list[i]
            i += 1
            k += 1
        while j < len(right_list):
            elements[k] = right_list[j]
            j += 1
            k += 1


def display_elements(elements):
    for i in elements:
        print(i, end=" ")
    print("")


def main():
    elements = [20, 12, 8, 10, 29, 7, 32, 23]
    merge_sort(elements)
    display_elements(elements)


if __name__ == "__main__":
    main()

File: C++/sorting/merge_sort.cpp
#include <iostream>
using namespace std;

void display_elements(int *a, int len)
{
    for (int i = 0; i < len; i++)
    {
        cout << a[i] << " ";
    }
    cout << endl;
}

// int count_length(int *a){
//     int count = 0;
//     for (int i = 0; a[i] != '\0'; i++)
//     {
//         count++;
//     }
//     return count;

// }

void merge_sort(int *arr, int len)
{ // 12 8 20 7 10

    if (len > 1)
    {
        int size_of_left_array = len / 2;                   // 1
        int size_of_right_array = len - size_of_left_array; // 1
        int left_array[size_of_left_array];
        for (int i = 0; i < size_of_left_array; i++)
        {
            left_array[i] = arr[i];
        }
        int right_array[size_of_right_array];
        for (int i = 0; i < size_of_right_array; i++)
        {
            right_array[i] = arr[size_of_left_array + i];
        }

        merge_sort(left_array, size_of_left_array);
        merge_sort(right_array, size_of_right_array);

        // now merge two sub array in ascending order

        int i = 0;
        int j = 0;
        int k = 0;

        while (i < size_of_left_array && j < size_of_right_array)
        {

            if (left_array[i] < right_array[j])
            {
                arr[k] = left_array[i];
                k++;
                i++;
            }
            else
            {
                arr[k] = right_array[j];
                k++;
                j++;
            }
        }

        while (i < size_of_left_array)
        {
            arr[k] = left_array[i];
            k++;
            i++;
        }
        while (j < size_of_right_array)
        {
            arr[k] = right_array[j];
            k++;
            j++;
        }
    }
}

int main()
{

    int a[] = {90,50,30,20,19,16,15,10,3,2};
    int len = 10;
    merge_sort(a, len);
    display_elements(a, len);

    return 0;
}